### PR TITLE
Add support for lock statements

### LINKED ISSUE #359: Add support for isolated functions
### Description

Isolated functions them self don't have any runtime semantics (potential optimization opportunities but no feature requirements). All we need to support is validate function to be isolated in frontend

### Describe your problem(s)

_No response_

### Describe your solution(s)

_No response_

### Related area

Compilation

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### LINKED ISSUE #422: Add support for lock statements
## Summary

Lock statements (`lock { ... }`) are not yet supported in the compiler.

## Background

This issue was raised in the context of PR #388 (Add support for isolated objects and isolated methods) and the review comment at https://github.com/ballerina-platform/ballerina-lang-go/pull/388#discussion_r3179233225.

## What needs to be done

1. Add parsing and semantic support for `lock` statements.
2. Fix `validateIsolatedClassMutableFieldAccess` in `semantics/semantic_analyzer.go` to track lock-nesting state during the AST walk and only emit the `"mutable field access within isolated object must be inside a lock statement"` diagnostic when the mutable `self` field access occurs **outside** a lock block. Currently the walk never checks whether the access is under a `lock` statement, so it incorrectly rejects valid patterns such as:

```ballerina
isolated class Counter {
    private int count = 0;
    function increment() {
        lock {
            self.count = self.count + 1; // should be allowed
        }
    }
}
```

## References

- PR: https://github.com/ballerina-platform/ballerina-lang-go/pull/388
- Review comment: https://github.com/ballerina-platform/ballerina-lang-go/pull/388#discussion_r3179233225
- Requested by: @heshanpadmasiri

### PR DESCRIPTION
**Pull request by heshanpadmasiri**
## Purpose
$subject
Resolves #422 and #428 
Depends on #424 

## Summary by CodeRabbit

* **New Features**
  * Added language-level lock statement with end-to-end support: parsing, printing, compilation, and emitted lock-start/lock-end in compiled output; proper handling across returns, breaks/continues and panics.

* **Tests**
  * Added a comprehensive test suite covering lock semantics, valid/invalid usage, nested/closure interactions, panic/release behavior, class/field rules, and control-flow edge cases.

### CODE REVIEW

File: corpus/bal/subset8/08-lock/lock-isolated-const-init-1-e.bal [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+isolated int seed = 7;
+
+public function main() {
+    int result = 0;
+    lock {
+        var f = function(int n = seed) returns int { // @error
+            return n;
+        };
+    }
+    io:println(result); // @output 7
```
[the remark this answers is not in the document]
Comment (heshanpadmasiri) [reply corpus/bal/subset8/08-lock/lock-isolated-const-init-1-e.bal]: Fixed (removed output marker this test shouldn't run -e)

File: bir/bir_gen.go [hunk 2]
```diff
@@ -482,19 +524,23 @@ func memberAccessInstructionKinds(tyCtx semtypes.Context, containerType semtypes
 }
 
 func continueStatement(ctx *stmtContext, curBB *BIRBasicBlock, stmt *ast.BLangContinue) statementEffect {
-	curBB.Instructions = append(curBB.Instructions, &PopScopeFrame{BIRInstructionBase: BIRInstructionBase{BIRNodeBase: BIRNodeBase{Pos: ctx.loc(stmt.GetPosition())}}})
+	pos := ctx.loc(stmt.GetPosition())
+	curBB = emitLockEndBeforeAbruptExit(ctx, curBB, pos)
```
Comment (gimantha) [inline bir/bir_gen.go]: Can you check this code against your PR?

```bal
ballerina/io;

isolated int counter = 0;

public function main() {
    lock {
        int i = 0;
        while i < 2 {
            i += 1;
            counter += 1;
            continue;
        }
    }

    lock {
        io:println(counter);
    }
}
```
If the loop is inside the lock statement and the continue statement should not release the lock.

Same thing goes with the break Statement as well

Comment (heshanpadmasiri) [reply bir/bir_gen.go]: Fixed with 99deb94

File: runtime/internal/locks/locks.go [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+	"math"
+	"sync"
+)
+
+// ReentrantMutex is a strand-aware re-entrant mutex. The same strand may
+// acquire the mutex any number of times; each acquisition must be paired
+// with a release before another strand can take ownership.
+type ReentrantMutex struct {
+	mu    sync.Mutex
+	cond  *sync.Cond
```
Comment (warunalakshitha) [inline runtime/internal/locks/locks.go:47]: Shall we add a comment describing usage of mu and cond

**Comment by warunalakshitha** [comment]
LGTM except minor changes requested.

**Review by gimantha — DISMISSED** [review]
LGTM

**Review by warunalakshitha — APPROVED** [review]
